crates/bevy_ecs/src/reflect/entity_commands.rs

```diff
@@ -182,113 +204,55 @@ impl<'w, 's, 'a> ReflectCommandExt for EntityCommands<'w, 's, 'a> {
     }
 }
 
-/// Helper function to add a reflect component to a given entity
-fn insert_reflect(
-    world: &mut World,
-    entity: Entity,
-    type_registry: &TypeRegistry,
-    component: Box<dyn Reflect>,
-) {
-    let type_info = component.type_name();
-    let Some(mut entity) = world.get_entity_mut(entity) else {
-        panic!("error[B0003]: Could not insert a reflected component (of type {}) for entity {entity:?} because it doesn't exist in this World.", component.type_name());
-    };
-    let Some(type_registration) = type_registry.get_with_name(type_info) else {
-        panic!("Could not get type registration (for component type {}) because it doesn't exist in the TypeRegistry.", component.type_name());
-    };
-    let Some(reflect_component) = type_registration.data::<ReflectComponent>() else {
-        panic!("Could not get ReflectComponent data (for component type {}) because it doesn't exist in this TypeRegistration.", component.type_name());
-    };
-    reflect_component.insert(&mut entity, &*component);
-}
-
-/// A [`Command`] that adds the boxed reflect component to an entity using the data in
-/// [`AppTypeRegistry`].
-///
-/// See [`ReflectCommandExt::insert_reflect`] for details.
-pub struct InsertReflect {
-    /// The entity on which the component will be inserted.
-    pub entity: Entity,
-    /// The reflect [Component](crate::component::Component) that will be added to the entity.
-    pub component: Box<dyn Reflect>,
-}
-
-impl Command for InsertReflect {
-    fn apply(self, world: &mut World) {
-        let registry = world.get_resource::<AppTypeRegistry>().unwrap().clone();
-        insert_reflect(world, self.entity, &registry.read(), self.component);
-    }
-}
-
 /// A [`Command`] that adds the boxed reflect component to an entity using the data in the provided
-/// [`Resource`] that implements [`AsRef<TypeRegistry>`].
+/// [`Resource`] that implements [`ReadTypeRegistry`].
 ///
 /// See [`ReflectCommandExt::insert_reflect_with_registry`] for details.
-pub struct InsertReflectWithRegistry<T: Resource + AsRef<TypeRegistry>> {
+pub struct InsertReflectWithRegistry<T: ReadTypeRegistry> {
     /// The entity on which the component will be inserted.
     pub entity: Entity,
     pub _t: PhantomData<T>,
     /// The reflect [Component](crate::component::Component) that will be added to the entity.
     pub component: Box<dyn Reflect>,
 }
 
-impl<T: Resource + AsRef<TypeRegistry>> Command for InsertReflectWithRegistry<T> {
+impl<T: ReadTypeRegistry> Command for InsertReflectWithRegistry<T> {
     fn apply(self, world: &mut World) {
         world.resource_scope(|world, registry: Mut<T>| {
```

Comment:
Note that previously `InsertReflect` (and `RemoveReflect` too) didn't use `resource_scope`, but instead `.clone()` the `AppTypeRegistry`. This meant that it was still available in the world when `reflect_component.insert` was called, but now it is no longer available. This could also be considered a breaking change considering custom implementations of `ReflectComponent` may be relying on that (#9623 for example relies on that).

A different approach could be using `get_resource` for getting the registry and then the `ReflectComponent`, then cloning the `ReflectComponent` (which is cheap, it's just a couple of function pointers, I think it could even be made `Copy`) to end the borrow of the world. Then you will be able to call `reflect_component.insert(&mut entity, &*component)` without borrow issues.

Reply:
The change was added after 0.11.2 I recall we decided to avoid adding to the breaking change section PRs that breaks APIs that were only introduced in the new release.

I like the suggestion though. I think it solves a real problem.